pkg/drawio/layoutMatrix.go

```diff
@@ -0,0 +1,175 @@
+package drawio
+
+const (
+	minMatrixSize = 30
+)
+
+type layer struct {
+	thickness int
+	distance  int
+	matrix    *layoutMatrix
+	index     int
+}
+
+func newLayer(matrix *layoutMatrix, index int) layer {
+	return layer{matrix: matrix, index: index}
+}
+
+type row struct {
+	layer
+}
+
+func newRow(matrix *layoutMatrix, index int) *row { return &row{newLayer(matrix, index)} }
+func (r *row) setHight(hight int)                 { r.thickness = hight }
+func (r *row) setY(y int)                         { r.distance = y }
+func (r *row) hight() int                         { return r.thickness }
+func (r *row) y() int                             { return r.distance }
+
+type col struct {
+	layer
+}
+
+func newCol(matrix *layoutMatrix, index int) *col { return &col{newLayer(matrix, index)} }
+func (c *col) setWidth(width int)                 { c.thickness = width }
+func (c *col) setX(y int)                         { c.distance = y }
+func (c *col) width() int                         { return c.thickness }
+func (c *col) x() int                             { return c.distance }
+
+/////////////////////////////////////////////////////////////////////////////////
+
+type Location struct {
+	firstRow *row
+	lastRow  *row
+	firstCol *col
+	lastCol  *col
+	xOffset  int
+	yOffset  int
+}
+
+func newLocation(firstRow, lastRow *row, firstCol, lastCol *col) *Location {
+	return &Location{firstRow, lastRow, firstCol, lastCol, 0, 0}
+}
+func newCellLocation(firstRow *row, firstCol *col) *Location {
+	return newLocation(firstRow, firstRow, firstCol, firstCol)
+}
+func (l *Location) nextRow() *row {
+	return l.lastRow.matrix.rows[l.lastRow.index+1]
+}
+func (l *Location) nextCol() *col {
+	return l.lastCol.matrix.cols[l.lastCol.index+1]
+}
+func (l *Location) prevRow() *row {
+	return l.firstRow.matrix.rows[l.firstRow.index-1]
+}
+func (l *Location) prevCol() *col {
+	return l.firstCol.matrix.cols[l.firstCol.index-1]
+}
+func (l *Location) copy() *Location {
+	nl := *l
+	return &nl
+}
+
```

Comment:
I'm not sure how safe this is - it's not really a deep copy. The firstRow, lastRow, firstCol, lastCol pointers will just be copied and refer to the existing objects.

Reply:
this exactly what I need.
to create a new location object, and copy its pointers 
of course I do not want that a new row/col will be created when creating a new location object

Reply:
Then what you get is Go's native semantics for assignment of objects (your code takes and returns pointers, so it's a bit different).
I recommend adding a comment: "Note: this is not a deep copy".

Reply:
done